File: AstroSleep-Android/app/src/main/java/com/astrosleep/app/core/engine/TagAffinityTables.kt
package com.astrosleep.app.core.engine

import com.astrosleep.app.core.model.Element
import com.astrosleep.app.core.model.MoonPhase
import com.astrosleep.app.core.model.Planet
import com.astrosleep.app.core.model.Sign

/**
 * Giant preference maps: chart symbols → sonic tag affinities.
 * Strengths are relative; TagEngine normalizes contribution at score time.
 */
object TagAffinityTables {

    /** Pool for per-user signature tag injection (deterministic from fingerprint). */
    val signaturePool: List<String> = listOf(
        "water", "air", "fire", "earth", "mechanical", "organic", "electrical", "cosmic",
        "steady", "pulse", "irregular", "chaotic", "rhythmic", "arrhythmic",
        "sub", "deep", "mid", "bright", "full", "ultrasonic",
        "nature", "domestic", "abstract", "urban", "industrial", "spiritual",
        "ethereal", "light", "medium", "heavy", "massive",
        "smooth", "rough", "crystalline", "diffuse", "granular", "glassy", "metallic",
        "static", "flowing", "surging", "swirling", "oscillating", "drifting", "pulsing",
        "vacuum", "sparse", "moderate", "dense", "saturated",
        "cold", "cool", "neutral", "warm", "hot",
        "active", "receptive", "balanced",
        "solar", "lunar", "stellar", "planetary", "void",
        "maiden", "mother", "crone", "hero", "mentor", "shadow", "trickster",
    )

    fun signTagPreferences(sign: Sign): Map<String, Double> = when (sign) {
        Sign.ARIES -> mapOf(
            "fire" to 1.4, "active" to 1.3, "surging" to 1.2, "hot" to 1.1,
            "hero" to 1.2, "rough" to 0.9, "pulse" to 1.0, "solar" to 0.8,
        )
        Sign.TAURUS -> mapOf(
            "earth" to 1.5, "steady" to 1.4, "heavy" to 1.2, "smooth" to 1.1,
            "mother" to 1.0, "organic" to 1.1, "static" to 0.9, "warm" to 0.8,
        )
        Sign.GEMINI -> mapOf(
            "air" to 1.4, "bright" to 1.3, "crystalline" to 1.2, "arrhythmic" to 1.0,
            "trickster" to 1.2, "electrical" to 1.0, "pulse" to 0.9, "light" to 0.9,
        )
        Sign.CANCER -> mapOf(
            "water" to 1.6, "lunar" to 1.5, "mother" to 1.4, "flowing" to 1.2,
            "receptive" to 1.3, "domestic" to 1.0, "cool" to 0.9, "dense" to 0.8,
        )
        Sign.LEO -> mapOf(
            "fire" to 1.4, "solar" to 1.5, "hero" to 1.3, "warm" to 1.2,
            "full" to 1.0, "active" to 1.1, "surging" to 0.9, "nature" to 0.7,
        )
        Sign.VIRGO -> mapOf(
            "earth" to 1.3, "granular" to 1.3, "moderate" to 1.1, "mentor" to 1.2,
            "steady" to 1.1, "mechanical" to 0.9, "mid" to 1.0, "organic" to 0.8,
        )
        Sign.LIBRA -> mapOf(
            "air" to 1.3, "balanced" to 1.5, "smooth" to 1.3, "maiden" to 1.1,
            "harmonic" to 0.0, "bright" to 1.0, "drifting" to 1.0, "spiritual" to 0.9,
        )
        Sign.SCORPIO -> mapOf(
            "water" to 1.4, "shadow" to 1.6, "dense" to 1.3, "deep" to 1.4,
            "heavy" to 1.1, "receptive" to 1.0, "void" to 0.9, "rough" to 0.8,
        )
        Sign.OPHIUCHUS -> mapOf(
            "water" to 1.2, "cosmic" to 1.4, "shadow" to 1.3, "mentor" to 1.2,
            "spiritual" to 1.3, "oscillating" to 1.1, "planetary" to 1.0, "crystalline" to 0.9,
        )
        Sign.SAGITTARIUS -> mapOf(
            "fire" to 1.3, "cosmic" to 1.2, "drifting" to 1.2, "hero" to 1.1,
            "bright" to 1.0, "nature" to 1.1, "active" to 0.9, "stellar" to 1.0,
        )
        Sign.CAPRICORN -> mapOf(
            "earth" to 1.5, "mechanical" to 1.2, "heavy" to 1.3, "crone" to 1.2,
            "static" to 1.1, "cold" to 1.0, "industrial" to 1.0, "steady" to 1.2,
        )
        Sign.AQUARIUS -> mapOf(
            "air" to 1.4, "electrical" to 1.5, "abstract" to 1.3, "ultrasonic" to 1.2,
            "trickster" to 1.1, "stellar" to 1.2, "vacuum" to 1.0, "glassy" to 1.0,
        )
        Sign.PISCES -> mapOf(
            "water" to 1.6, "ethereal" to 1.4, "diffuse" to 1.3, "drifting" to 1.3,
            "spiritual" to 1.4, "lunar" to 1.2, "receptive" to 1.2, "shadow" to 0.9,
        )
    }

    fun planetTagPreferences(planet: Planet): Map<String, Double> = when (planet) {
        Planet.SUN -> mapOf("solar" to 1.4, "warm" to 1.1, "active" to 1.0, "fire" to 0.9, "hero" to 0.8)
        Planet.MOON -> mapOf("lunar" to 1.6, "water" to 1.2, "receptive" to 1.2, "mother" to 1.0, "flowing" to 0.9)
        Planet.MERCURY -> mapOf("bright" to 1.1, "electrical" to 1.0, "pulse" to 1.0, "air" to 0.9, "trickster" to 0.8)
        Planet.VENUS -> mapOf("smooth" to 1.3, "warm" to 1.1, "balanced" to 1.0, "organic" to 0.9, "maiden" to 0.8)
        Planet.MARS -> mapOf("fire" to 1.3, "rough" to 1.2, "active" to 1.2, "surging" to 1.1, "hot" to 1.0)
        Planet.JUPITER -> mapOf("full" to 1.1, "cosmic" to 1.0, "expansive" to 0.0, "nature" to 0.9, "mentor" to 0.9)
        Planet.SATURN -> mapOf("earth" to 1.2, "heavy" to 1.3, "mechanical" to 1.1, "static" to 1.1, "crone" to 1.0, "cold" to 0.9)
        Planet.URANUS -> mapOf("electrical" to 1.5, "chaotic" to 1.2, "ultrasonic" to 1.2, "abstract" to 1.1, "air" to 0.9)
        Planet.NEPTUNE -> mapOf("water" to 1.3, "ethereal" to 1.4, "diffuse" to 1.3, "spiritual" to 1.2, "drifting" to 1.2, "void" to 0.8)
        Planet.PLUTO -> mapOf("shadow" to 1.5, "dense" to 1.3, "deep" to 1.3, "heavy" to 1.1, "water" to 0.9)
        Planet.CHIRON -> mapOf("mentor" to 1.3, "spiritual" to 1.1, "organic" to 0.9, "mid" to 0.8, "cosmic" to 0.8)
        Planet.LILITH -> mapOf("shadow" to 1.4, "void" to 1.2, "irregular" to 1.0, "rough" to 0.9, "lunar" to 0.8)
        Planet.NORTH_NODE -> mapOf("stellar" to 1.1, "active" to 0.9, "cosmic" to 1.0, "drifting" to 0.8)
        Planet.SOUTH_NODE -> mapOf("void" to 1.1, "crone" to 0.9, "static" to 0.8, "shadow" to 0.8)
    }

    fun elementTagPreferences(element: Element): Map<String, Double> = when (element) {
        Element.FIRE -> mapOf(
            "fire" to 1.5, "hot" to 1.3, "active" to 1.2, "surging" to 1.1,
            "rough" to 0.9, "solar" to 1.0, "hero" to 1.0, "pulse" to 0.8,
        )
        Element.EARTH -> mapOf(
            "earth" to 1.5, "heavy" to 1.3, "steady" to 1.3, "dense" to 1.2,
            "static" to 1.1, "organic" to 1.0, "mechanical" to 0.9, "smooth" to 0.8,
        )
        Element.AIR -> mapOf(
            "air" to 1.5, "bright" to 1.3, "light" to 1.2, "crystalline" to 1.2,
            "electrical" to 1.1, "drifting" to 1.1, "sparse" to 1.0, "trickster" to 0.9,
        )
        Element.WATER -> mapOf(
            "water" to 1.6, "flowing" to 1.4, "cool" to 1.2, "receptive" to 1.3,
            "lunar" to 1.2, "dense" to 1.0, "mother" to 1.1, "diffuse" to 1.0,
        )
    }

    fun moonPhaseTagPreferences(phase: MoonPhase): Map<String, Double> = when (phase) {
        MoonPhase.NEW_MOON -> mapOf(
            "void" to 1.4, "sparse" to 1.3, "ethereal" to 1.2, "sub" to 1.1,
            "receptive" to 1.2, "shadow" to 1.0, "cold" to 0.9,
        )
        MoonPhase.WAXING_CRESCENT -> mapOf(
            "light" to 1.2, "maiden" to 1.2, "pulse" to 1.1, "bright" to 1.0,
            "active" to 0.9, "air" to 0.8,
        )
        MoonPhase.FIRST_QUARTER -> mapOf(
            "active" to 1.3, "pulse" to 1.2, "fire" to 1.0, "rough" to 0.9,
            "hero" to 1.0, "mid" to 0.8,
        )
        MoonPhase.WAXING_GIBBOUS -> mapOf(
            "dense" to 1.1, "full" to 1.1, "organic" to 1.0, "warm" to 1.0,
            "flowing" to 0.9, "balanced" to 0.9,
        )
        MoonPhase.FULL_MOON -> mapOf(
            "lunar" to 1.5, "bright" to 1.2, "saturated" to 1.2, "water" to 1.1,
            "air" to 1.0, "full" to 1.1, "crystalline" to 1.0,
        )
        MoonPhase.WANING_GIBBOUS -> mapOf(
            "diffuse" to 1.2, "mentor" to 1.1, "drifting" to 1.1, "cool" to 1.0,
            "spiritual" to 1.0, "moderate" to 0.9,
        )
        MoonPhase.LAST_QUARTER -> mapOf(
            "earth" to 1.2, "crone" to 1.2, "steady" to 1.1, "heavy" to 1.0,
            "static" to 0.9, "mechanical" to 0.8,
        )
        MoonPhase.WANING_CRESCENT -> mapOf(
            "water" to 1.3, "ethereal" to 1.3, "receptive" to 1.3, "void" to 1.1,
            "drifting" to 1.2, "sparse" to 1.0, "shadow" to 1.0, "cool" to 1.1,
        )
    }

    /**
     * Which tags "play well" for each stack role — used when picking layers.
     * Higher score = better fit for that seat in the mix.
     */
    fun roleFit(role: StackRole, tags: com.astrosleep.app.core.model.SoundTags): Double {
        var s = 0.0
        fun hit(cond: Boolean, w: Double) { if (cond) s += w }
        when (role) {
            StackRole.BEDROCK -> {
                hit(tags.register in setOf("sub", "deep"), 2.2)
                hit(tags.weight in setOf("heavy", "massive"), 1.8)
                hit(tags.density in setOf("dense", "saturated"), 1.4)
                hit(tags.domain in setOf("earth", "mechanical", "water"), 1.2)
                hit(tags.motion == "static" || tags.rhythm == "steady", 1.0)
            }
            StackRole.FOUNDATION -> {
                hit(tags.domain in setOf("water", "earth", "organic", "air"), 1.6)
                hit(tags.weight in setOf("medium", "heavy"), 1.2)
                hit(tags.density in setOf("moderate", "dense"), 1.0)
                hit(tags.context == "nature" || tags.context == "domestic", 0.8)
            }
            StackRole.FLOW -> {
                hit(tags.motion in setOf("flowing", "drifting", "swirling"), 2.0)
                hit(tags.domain == "water" || tags.domain == "air", 1.4)
                hit(tags.rhythm in setOf("irregular", "arrhythmic", "pulse"), 1.0)
            }
            StackRole.TEXTURE -> {
                hit(tags.texture in setOf("granular", "rough", "crystalline", "metallic", "glassy"), 2.0)
                hit(tags.domain in setOf("mechanical", "electrical", "organic"), 1.2)
                hit(tags.register in setOf("mid", "bright"), 0.9)
            }
            StackRole.VEIL -> {
                hit(tags.weight in setOf("ethereal", "light"), 2.0)
                hit(tags.texture in setOf("diffuse", "smooth", "crystalline"), 1.5)
                hit(tags.density in setOf("sparse", "vacuum", "moderate"), 1.2)
                hit(tags.celestial in setOf("lunar", "stellar", "void"), 1.0)
            }
            StackRole.SPARK -> {
                hit(tags.register in setOf("bright", "ultrasonic", "full"), 2.0)
                hit(tags.temperature in setOf("warm", "hot"), 1.2)
                hit(tags.domain in setOf("fire", "electrical", "cosmic"), 1.3)
                hit(tags.polarity == "active", 1.0)
            }
            StackRole.ACCENT -> {
                hit(tags.archetype in setOf("trickster", "hero", "shadow", "maiden"), 1.5)
                hit(tags.context in setOf("urban", "industrial", "spiritual", "abstract"), 1.2)
                hit(tags.celestial in setOf("planetary", "solar", "stellar"), 1.0)
                hit(tags.isNewLike(), 0.3)
            }
        }
        return s
    }

    private fun com.astrosleep.app.core.model.SoundTags.isNewLike(): Boolean =
        domain in setOf("electrical", "cosmic") || texture in setOf("glassy", "metallic")
}

File: AstroSleep-iOS/AstroSleep/Core/Engine/TagAffinityTables.swift
import Foundation

// MARK: - Tag affinity tables (parity with Android TagAffinityTables.kt)

/// Chart symbols → sonic tag affinities. Strengths are relative;
/// TagEngine normalizes contribution at score time.
enum TagAffinityTables {

    /// Pool for per-user signature tag injection (deterministic from fingerprint).
    static let signaturePool: [String] = [
        "water", "air", "fire", "earth", "mechanical", "organic", "electrical", "cosmic",
        "steady", "pulse", "irregular", "chaotic", "rhythmic", "arrhythmic",
        "sub", "deep", "mid", "bright", "full", "ultrasonic",
        "nature", "domestic", "abstract", "urban", "industrial", "spiritual",
        "ethereal", "light", "medium", "heavy", "massive",
        "smooth", "rough", "crystalline", "diffuse", "granular", "glassy", "metallic",
        "static", "flowing", "surging", "swirling", "oscillating", "drifting", "pulsing",
        "vacuum", "sparse", "moderate", "dense", "saturated",
        "cold", "cool", "neutral", "warm", "hot",
        "active", "receptive", "balanced",
        "solar", "lunar", "stellar", "planetary", "void",
        "maiden", "mother", "crone", "hero", "mentor", "shadow", "trickster"
    ]

    static func signTagPreferences(_ sign: Sign) -> [String: Double] {
        switch sign {
        case .aries:
            return ["fire": 1.4, "active": 1.3, "surging": 1.2, "hot": 1.1,
                    "hero": 1.2, "rough": 0.9, "pulse": 1.0, "solar": 0.8]
        case .taurus:
            return ["earth": 1.5, "steady": 1.4, "heavy": 1.2, "smooth": 1.1,
                    "mother": 1.0, "organic": 1.1, "static": 0.9, "warm": 0.8]
        case .gemini:
            return ["air": 1.4, "bright": 1.3, "crystalline": 1.2, "arrhythmic": 1.0,
                    "trickster": 1.2, "electrical": 1.0, "pulse": 0.9, "light": 0.9]
        case .cancer:
            return ["water": 1.6, "lunar": 1.5, "mother": 1.4, "flowing": 1.2,
                    "receptive": 1.3, "domestic": 1.0, "cool": 0.9, "dense": 0.8]
        case .leo:
            return ["fire": 1.4, "solar": 1.5, "hero": 1.3, "warm": 1.2,
                    "full": 1.0, "active": 1.1, "surging": 0.9, "nature": 0.7]
        case .virgo:
            return ["earth": 1.3, "granular": 1.3, "moderate": 1.1, "mentor": 1.2,
                    "steady": 1.1, "mechanical": 0.9, "mid": 1.0, "organic": 0.8]
        case .libra:
            return ["air": 1.3, "balanced": 1.5, "smooth": 1.3, "maiden": 1.1,
                    "bright": 1.0, "drifting": 1.0, "spiritual": 0.9]
        case .scorpio:
            return ["water": 1.4, "shadow": 1.6, "dense": 1.3, "deep": 1.4,
                    "heavy": 1.1, "receptive": 1.0, "void": 0.9, "rough": 0.8]
        case .ophiuchus:
            return ["water": 1.2, "cosmic": 1.4, "shadow": 1.3, "mentor": 1.2,
                    "spiritual": 1.3, "oscillating": 1.1, "planetary": 1.0, "crystalline": 0.9]
        case .sagittarius:
            return ["fire": 1.3, "cosmic": 1.2, "drifting": 1.2, "hero": 1.1,
                    "bright": 1.0, "nature": 1.1, "active": 0.9, "stellar": 1.0]
        case .capricorn:
            return ["earth": 1.5, "mechanical": 1.2, "heavy": 1.3, "crone": 1.2,
                    "static": 1.1, "cold": 1.0, "industrial": 1.0, "steady": 1.2]
        case .aquarius:
            return ["air": 1.4, "electrical": 1.5, "abstract": 1.3, "ultrasonic": 1.2,
                    "trickster": 1.1, "stellar": 1.2, "vacuum": 1.0, "glassy": 1.0]
        case .pisces:
            return ["water": 1.6, "ethereal": 1.4, "diffuse": 1.3, "drifting": 1.3,
                    "spiritual": 1.4, "lunar": 1.2, "receptive": 1.2, "shadow": 0.9]
        }
    }

    static func planetTagPreferences(_ planet: Planet) -> [String: Double] {
        switch planet {
        case .sun:
            return ["solar": 1.4, "warm": 1.1, "active": 1.0, "fire": 0.9, "hero": 0.8]
        case .moon:
            return ["lunar": 1.6, "water": 1.2, "receptive": 1.2, "mother": 1.0, "flowing": 0.9]
        case .mercury:
            return ["bright": 1.1, "electrical": 1.0, "pulse": 1.0, "air": 0.9, "trickster": 0.8]
        case .venus:
            return ["smooth": 1.3, "warm": 1.1, "balanced": 1.0, "organic": 0.9, "maiden": 0.8]
        case .mars:
            return ["fire": 1.3, "rough": 1.2, "active": 1.2, "surging": 1.1, "hot": 1.0]
        case .jupiter:
            return ["full": 1.1, "cosmic": 1.0, "nature": 0.9, "mentor": 0.9]
        case .saturn:
            return ["earth": 1.2, "heavy": 1.3, "mechanical": 1.1, "static": 1.1, "crone": 1.0, "cold": 0.9]
        case .uranus:
            return ["electrical": 1.5, "chaotic": 1.2, "ultrasonic": 1.2, "abstract": 1.1, "air": 0.9]
        case .neptune:
            return ["water": 1.3, "ethereal": 1.4, "diffuse": 1.3, "spiritual": 1.2, "drifting": 1.2, "void": 0.8]
        case .pluto:
            return ["shadow": 1.5, "dense": 1.3, "deep": 1.3, "heavy": 1.1, "water": 0.9]
        case .chiron:
            return ["mentor": 1.3, "spiritual": 1.1, "organic": 0.9, "mid": 0.8, "cosmic": 0.8]
        case .lilith:
            return ["shadow": 1.4, "void": 1.2, "irregular": 1.0, "rough": 0.9, "lunar": 0.8]
        case .northNode:
            return ["stellar": 1.1, "active": 0.9, "cosmic": 1.0, "drifting": 0.8]
        case .southNode:
            return ["void": 1.1, "crone": 0.9, "static": 0.8, "shadow": 0.8]
        }
    }

    static func elementTagPreferences(_ element: Element) -> [String: Double] {
        switch element {
        case .fire:
            return ["fire": 1.5, "hot": 1.3, "active": 1.2, "surging": 1.1,
                    "rough": 0.9, "solar": 1.0, "hero": 1.0, "pulse": 0.8]
        case .earth:
            return ["earth": 1.5, "heavy": 1.3, "steady": 1.3, "dense": 1.2,
                    "static": 1.1, "organic": 1.0, "mechanical": 0.9, "smooth": 0.8]
        case .air:
            return ["air": 1.5, "bright": 1.3, "light": 1.2, "crystalline": 1.2,
                    "electrical": 1.1, "drifting": 1.1, "sparse": 1.0, "trickster": 0.9]
        case .water:
            return ["water": 1.6, "flowing": 1.4, "cool": 1.2, "receptive": 1.3,
                    "lunar": 1.2, "dense": 1.0, "mother": 1.1, "diffuse": 1.0]
        }
    }

    static func moonPhaseTagPreferences(_ phase: MoonPhase) -> [String: Double] {
        switch phase {
        case .newMoon:
            return ["void": 1.4, "sparse": 1.3, "ethereal": 1.2, "sub": 1.1,
                    "receptive": 1.2, "shadow": 1.0, "cold": 0.9]
        case .waxingCrescent:
            return ["light": 1.2, "maiden": 1.2, "pulse": 1.1, "bright": 1.0,
                    "active": 0.9, "air": 0.8]
        case .firstQuarter:
            return ["active": 1.3, "pulse": 1.2, "fire": 1.0, "rough": 0.9,
                    "hero": 1.0, "mid": 0.8]
        case .waxingGibbous:
            return ["dense": 1.1, "full": 1.1, "organic": 1.0, "warm": 1.0,
                    "flowing": 0.9, "balanced": 0.9]
        case .fullMoon:
            return ["lunar": 1.5, "bright": 1.2, "saturated": 1.2, "water": 1.1,
                    "air": 1.0, "full": 1.1, "crystalline": 1.0]
        case .waningGibbous:
            return ["diffuse": 1.2, "mentor": 1.1, "drifting": 1.1, "cool": 1.0,
                    "spiritual": 1.0, "moderate": 0.9]
        case .lastQuarter:
            return ["earth": 1.2, "crone": 1.2, "steady": 1.1, "heavy": 1.0,
                    "static": 0.9, "mechanical": 0.8]
        case .waningCrescent:
            return ["water": 1.3, "ethereal": 1.3, "receptive": 1.3, "void": 1.1,
                    "drifting": 1.2, "sparse": 1.0, "shadow": 1.0, "cool": 1.1]
        }
    }

    /// Which tags play well for each stack role — used when picking layers.
    static func roleFit(_ role: StackRole, tags: SoundTags) -> Double {
        var s = 0.0
        func hit(_ cond: Bool, _ w: Double) { if cond { s += w } }
        switch role {
        case .bedrock:
            hit(["sub", "deep"].contains(tags.register), 2.2)
            hit(["heavy", "massive"].contains(tags.weight), 1.8)
            hit(["dense", "saturated"].contains(tags.density), 1.4)
            hit(["earth", "mechanical", "water"].contains(tags.domain), 1.2)
            hit(tags.motion == "static" || tags.rhythm == "steady", 1.0)
        case .foundation:
            hit(["water", "earth", "organic", "air"].contains(tags.domain), 1.6)
            hit(["medium", "heavy"].contains(tags.weight), 1.2)
            hit(["moderate", "dense"].contains(tags.density), 1.0)
            hit(tags.context == "nature" || tags.context == "domestic", 0.8)
        case .flow:
            hit(["flowing", "drifting", "swirling"].contains(tags.motion), 2.0)
            hit(tags.domain == "water" || tags.domain == "air", 1.4)
            hit(["irregular", "arrhythmic", "pulse"].contains(tags.rhythm), 1.0)
        case .texture:
            hit(["granular", "rough", "crystalline", "metallic", "glassy"].contains(tags.texture), 2.0)
            hit(["mechanical", "electrical", "organic"].contains(tags.domain), 1.2)
            hit(["mid", "bright"].contains(tags.register), 0.9)
        case .veil:
            hit(["ethereal", "light"].contains(tags.weight), 2.0)
            hit(["diffuse", "smooth", "crystalline"].contains(tags.texture), 1.5)
            hit(["sparse", "vacuum", "moderate"].contains(tags.density), 1.2)
            hit(["lunar", "stellar", "void"].contains(tags.celestial), 1.0)
        case .spark:
            hit(["bright", "ultrasonic", "full"].contains(tags.register), 2.0)
            hit(["warm", "hot"].contains(tags.temperature), 1.2)
            hit(["fire", "electrical", "cosmic"].contains(tags.domain), 1.3)
            hit(tags.polarity == "active", 1.0)
        case .accent:
            hit(["trickster", "hero", "shadow", "maiden"].contains(tags.archetype), 1.5)
            hit(["urban", "industrial", "spiritual", "abstract"].contains(tags.context), 1.2)
            hit(["planetary", "solar", "stellar"].contains(tags.celestial), 1.0)
            hit(isNewLike(tags), 0.3)
        }
        return s
    }

    private static func isNewLike(_ tags: SoundTags) -> Bool {
        ["electrical", "cosmic"].contains(tags.domain)
            || ["glassy", "metallic"].contains(tags.texture)
    }
}
